### PR TITLE
ToolsPanel: Add subheadings and reset text to tools panel menu

### LINKED ISSUE #44096: ToolsPanel: Improve ellipsis menu with subheadings and separate reset button for default tools
The [ToolsPanel](https://wordpress.github.io/gutenberg/?path=/story/components-experimental-toolspanel--default) component comes with an ellipsis menu. This ellipsis menu serves two purposes:

1. To make available design tools that are occasionally but not always useful
2. To clear various such properties back to their default state

This panel is in part created to allow more design tools in a legible state, without encumbering the panel with repeated "Reset" buttons. In most cases, after all, you can select "Default" from a drop-down, or select and clear a line-height value, accomplishing the same. here's what it looks like today:

Typography panel today

For the cases where it still needs to be clearer how to reset a value back to its defaults, however, we can improve the clarity of the menu. 

Typography panel with tweaks

Shown left, a Typography panel with the default font size shown, but no value chosen.  Shown right, the panel with the additional non-default Font and Appearance tools shown, including a value picked for font size. The main changes are:

* Sections have subheadings delineating default tools (always shown in the panel) and additional tools.
* The default tool is no longer a full button that gets disabled when it has no set value, instead it gets an explicit "Reset" button on the right.
* Additional tool can be cleared and removed by unchecking, just as today.

Note that these improvements do not preclude additional enhancements, but they could be a starting point. See also #43082.

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Fixes: https://github.com/WordPress/gutenberg/issues/44096

## What?

- Adds subheadings to the `ToolsPanel` ellipsis menu
- Adds reset text within the default control menu items
- Updates `MenuItem` to allow more than just an icon or shortcut as a suffix

## Why?

Improve the `ToolsPanel` ellipsis menu UX.

## How?

- Updates the `MenuItem` to allow non icon or shortcut content (i.e. "reset" text)
- Leverages new `suffix` prop for a `MenuItem` to set reset text for default control items in the `ToolsPanel`
- Tweaks `ToolsPanel` styles to match the mockups in [#44096](https://github.com/WordPress/gutenberg/issues/44096)
- Updates `ToolsPanel` stories to provide better demonstration of the ellipsis menu for the default story

## Testing Instructions
1. Fire up Storybook and check out the ToolsPanel stories. Ensure the menu looks and behaves as per [#44096](https://github.com/WordPress/gutenberg/issues/44096)
2. Test out the various ToolsPanel uses in the editors e.g. the block support panels for typography, dimensions etc
3. Run `npm run test:unit packages/components/src/menu-item`

## Screenshots or screencast 

| Storybook | Editor |
|---|---|
| <video src="https://user-images.githubusercontent.com/60436221/190987707-4c52ad42-e6b9-4a2b-8f54-471c887aefd3.mp4" /> | <video src="https://user-images.githubusercontent.com/60436221/190987699-fb7cdea9-874c-4b2d-be38-fac384b130ff.mp4" /> |

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
@ciampo or @mirka Do you have any suggestions on a better approach to injecting "reset" text into the default control menu items for the `ToolsPanel`?

At present this PR will add a new `suffix` prop to the `MenuItem` component. We could avoid that if we render both the main label and reset text as children to the `MenuItem`. The downside is some pretty severe specificity and hacky CSS to make it align with the design mockup.

<details>
<summary>An alternative approach to updating `MenuItem`</summary>

The following patch was a result of initially hacking around exploring options for #44096. It can be applied to trunk and tested if so desired.

```diff
diff --git a/packages/components/src/tools-panel/stories/index.js b/packages/components/src/tools-panel/stories/index.js
index 803a01eb34..2a550b36cb 100644
--- a/packages/components/src/tools-panel/stories/index.js
+++ b/packages/components/src/tools-panel/stories/index.js
@@ -27,6 +27,7 @@ export default {
 
 export const _default = () => {
 	const [ height, setHeight ] = useState();
+	const [ maxHeight, setMaxHeight ] = useState();
 	const [ minHeight, setMinHeight ] = useState();
 	const [ width, setWidth ] = useState();
 
@@ -71,7 +72,6 @@ export const _default = () => {
 						hasValue={ () => !! minHeight }
 						label="Minimum height"
 						onDeselect={ () => setMinHeight( undefined ) }
-						isShownByDefault={ true }
 					>
 						<UnitControl
 							label="Minimum height"
@@ -79,6 +79,17 @@ export const _default = () => {
 							onChange={ ( next ) => setMinHeight( next ) }
 						/>
 					</ToolsPanelItem>
+					<ToolsPanelItem
+						hasValue={ () => !! maxHeight }
+						label="Maximum height"
+						onDeselect={ () => setMaxHeight( undefined ) }
+					>
+						<UnitControl
+							label="Maximum height"
+							value={ maxHeight }
+							onChange={ ( next ) => setMaxHeight( next ) }
+						/>
+					</ToolsPanelItem>
 				</ToolsPanel>
 			</Panel>
 		</PanelWrapperView>
diff --git a/packages/components/src/tools-panel/styles.ts b/packages/components/src/tools-panel/styles.ts
index 7fe2fc0a0a..9c82f740c0 100644
--- a/packages/components/src/tools-panel/styles.ts
+++ b/packages/components/src/tools-panel/styles.ts
@@ -12,7 +12,7 @@ import {
 	Wrapper as BaseControlWrapper,
 } from '../base-control/styles/base-control-styles';
 import { LabelWrapper } from '../input-control/styles/input-control-styles';
-import { COLORS, CONFIG } from '../utils';
+import { baseLabelTypography, COLORS, CONFIG } from '../utils';
 import { space } from '../ui/utils/space';
 
 const toolsPanelGrid = {
@@ -145,3 +145,32 @@ export const ToolsPanelItemPlaceholder = css`
 export const DropdownMenu = css`
 	min-width: 200px;
 `;
+
+// Following provides for rendering "reset" text within a MenuItem.
+export const DefaultControlsItem = css`
+	/**
+	 * This high level of specificity is needed to overcome styles for items
+	 * with no icon or shortcut.
+	 */
+	&&&&& > span {
+		display: flex;
+		/* item text plus icon min-width to match optional items. */
+		min-width: 208px;
+		justify-content: space-between;
+		padding-right: 0;
+	}
+
+	span > span:last-child {
+		flex: 0;
+		color: var( --wp-admin-theme-color-darker-10 );
+		${ baseLabelTypography }
+	}
+
+	&&[aria-disabled='true'] {
+		opacity: 1;
+
+		span > span:last-child {
+			opacity: 0.3;
+		}
+	}
+`;
diff --git a/packages/components/src/tools-panel/tools-panel-header/component.tsx b/packages/components/src/tools-panel/tools-panel-header/component.tsx
index 62a40a966d..414256f5ed 100644
--- a/packages/components/src/tools-panel/tools-panel-header/component.tsx
+++ b/packages/components/src/tools-panel/tools-panel-header/component.tsx
@@ -7,7 +7,7 @@ import type { ForwardedRef } from 'react';
  * WordPress dependencies
  */
 import { speak } from '@wordpress/a11y';
-import { check, reset, moreVertical, plus } from '@wordpress/icons';
+import { check, moreVertical, plus } from '@wordpress/icons';
 import { __, _x, sprintf } from '@wordpress/i18n';
 
 /**
@@ -26,6 +26,7 @@ import type {
 } from '../types';
 
 const DefaultControlsGroup = ( {
+	itemClassName,
 	items,
 	toggleItem,
 }: ToolsPanelControlsGroupProps ) => {
@@ -34,14 +35,14 @@ const DefaultControlsGroup = ( {
 	}
 
 	return (
-		<MenuGroup>
+		<MenuGroup label={ __( 'Defaults' ) }>
 			{ items.map( ( [ label, hasValue ] ) => {
 				if ( hasValue ) {
 					return (
 						<MenuItem
 							key={ label }
+							className={ itemClassName }
 							role="menuitem"
-							icon={ reset }
 							label={ sprintf(
 								// translators: %s: The name of the control being reset e.g. "Padding".
 								__( 'Reset %s' ),
@@ -59,7 +60,8 @@ const DefaultControlsGroup = ( {
 								);
 							} }
 						>
-							{ label }
+							<span>{ label }</span>
+							<span>{ __( 'Reset' ) }</span>
 						</MenuItem>
 					);
 				}
@@ -67,12 +69,13 @@ const DefaultControlsGroup = ( {
 				return (
 					<MenuItem
 						key={ label }
+						className={ itemClassName }
 						role="menuitemcheckbox"
-						icon={ check }
 						isSelected
 						aria-disabled
 					>
-						{ label }
+						<span>{ label }</span>
+						<span>{ __( 'Reset' ) }</span>
 					</MenuItem>
 				);
 			} ) }
@@ -89,7 +92,7 @@ const OptionalControlsGroup = ( {
 	}
 
 	return (
-		<MenuGroup>
+		<MenuGroup label={ __( 'Tools' ) }>
 			{ items.map( ( [ label, isSelected ] ) => {
 				const itemLabel = isSelected
 					? sprintf(
@@ -147,6 +150,7 @@ const ToolsPanelHeader = (
 ) => {
 	const {
 		areAllOptionalControlsHidden,
+		defaultControlsItemClassName,
 		dropdownMenuClassName,
 		hasMenuItems,
 		headingClassName,
@@ -197,6 +201,7 @@ const ToolsPanelHeader = (
 							<DefaultControlsGroup
 								items={ defaultItems }
 								toggleItem={ toggleItem }
+								itemClassName={ defaultControlsItemClassName }
 							/>
 							<OptionalControlsGroup
 								items={ optionalItems }
diff --git a/packages/components/src/tools-panel/tools-panel-header/hook.ts b/packages/components/src/tools-panel/tools-panel-header/hook.ts
index 586fd0e7c5..b44dc5bc9f 100644
--- a/packages/components/src/tools-panel/tools-panel-header/hook.ts
+++ b/packages/components/src/tools-panel/tools-panel-header/hook.ts
@@ -33,12 +33,17 @@ export function useToolsPanelHeader(
 		return cx( styles.ToolsPanelHeading );
 	}, [ cx ] );
 
+	const defaultControlsItemClassName = useMemo( () => {
+		return cx( styles.DefaultControlsItem );
+	}, [ cx ] );
+
 	const { menuItems, hasMenuItems, areAllOptionalControlsHidden } =
 		useToolsPanelContext();
 
 	return {
 		...otherProps,
 		areAllOptionalControlsHidden,
+		defaultControlsItemClassName,
 		dropdownMenuClassName,
 		hasMenuItems,
 		headingClassName,
diff --git a/packages/components/src/tools-panel/types.ts b/packages/components/src/tools-panel/types.ts
index 3290b9fe49..b95a50a4aa 100644
--- a/packages/components/src/tools-panel/types.ts
+++ b/packages/components/src/tools-panel/types.ts
@@ -147,6 +147,7 @@ export type ToolsPanelContext = {
 };
 
 export type ToolsPanelControlsGroupProps = {
+	itemClassName?: string;
 	items: [ string, boolean ][];
 	toggleItem: ( label: string ) => void;
 };

```

</details>

**Comment by jasmussen** [comment]
Here's the tools panel today:

before

Here's after:

after

I think it's a small but good improvement and I'd be happy to land this one to move things forward.

This was the case before as well, and remains the case here: focus is set on the first item in the menu, even if it's disabled. I understand that this is done to prevent a focus loss in the case where you _clear the font size which should render it disabled_, but it feels sub-optimal. This is something to think about.

**Review by jasmussen — APPROVED** [review]
Would love a quick gut check by @WordPress/gutenberg-design. Thanks for working on this!

**Comment by talldan** [comment]
Hmm, to me, this still looks like an enabled menu item and feels clickable (even though there's no hover style):
Screen Shot 2022-09-19 at 5 58 03 pm

It would be good to get other opinions though. I know in the issue using the same grey color as the Save button was mentioned as a potential option (https://github.com/WordPress/gutenberg/issues/44096#issuecomment-1247762119).

When using Voiceover and focusing the Font size item, the label is read as 'Font size RESET'. It might be worth making the 'RESET' text aria-hidden, as I think it may end up being confusing.

When the button is active it has its own aria-label ('Reset font size'), so it won't be an issue making the 'RESET' text aria-hidden.

File: packages/components/src/menu-item/index.js
Comment (mirka) [inline packages/components/src/menu-item/index.js:76]: Since this is a new API and we can kind of "start from scratch", what do you think about **not** rendering the `shortcut` and the `right` icon element if a `suffix` is used? I say this because I feel like we might want to eventually deprecate the `icon` and `shortcut` props here and unify everything into the `prefix`/`suffix` props. That would make things simpler for this component itself, and also better for API consistency across the package.

File: packages/components/src/tools-panel/styles.ts
Comment (mirka) [inline packages/components/src/tools-panel/styles.ts]: Since we are the ones providing the "Reset" span, would it be possible to avoid a `span:last-child` selector and target it more explicitly? It's hard to tell what `span:last-child` is, and it's a bit fragile.

File: packages/components/src/tools-panel/styles.ts
Comment (mirka) [inline packages/components/src/tools-panel/styles.ts]: What do you think about making this margin-left come as a part of the `suffix` prop?

File: packages/components/src/tools-panel/styles.ts
Comment (mirka) [inline packages/components/src/tools-panel/styles.ts]: Not super important, but I'm mildly cautious about using this mixin here because it's not the intended use case. What makes sense as a style change for the intended use case may not make sense in this menu item context. I'm fine with copy-pasting the actual styles if that makes more sense.

**Comment by aaronrobertshaw** [comment]
Appreciate the feedback @talldan 🙇 

> I know in the issue using the same grey color as the Save button was mentioned as a potential option (https://github.com/WordPress/gutenberg/issues/44096#issuecomment-1247762119).

I've tweaked the text color in the disabled state to match the Saved button as suggested. 

> When using Voiceover and focusing the Font size item, the label is read as 'Font size RESET'. It might be worth making the 'RESET' text aria-hidden, as I think it may end up being confusing.

Good catch! I've added aria-hidden to the reset text as well. It does read a little nicer 👍 



https://user-images.githubusercontent.com/60436221/191031639-8bd839ae-e3e3-47ba-8e54-b05a5faa5c6f.mp4

File: packages/components/src/menu-item/index.js
Comment (aaronrobertshaw) [reply packages/components/src/menu-item/index.js:76]: That sounds like a good plan to me 👍 

I've updated `MenuItem` according to your suggestion in https://github.com/WordPress/gutenberg/pull/44260/commits/ecffad03ca51b8f68f6592d8d972c5ef023af2c8

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: We certainly can. In the `span:last-child` approach I was hoping to avoid plumbing through multiple dynamic classnames. I've pushed a different solution using a `styled.span` in https://github.com/WordPress/gutenberg/pull/44260/commits/c8fa09dc204bdf95faa19f3581e3ab6a16f469bc. Is that any better?

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: By "come as part of the suffix prop", do you mean that the MenuItem component would apply a left margin here by default? I could see different suffixes desiring no margin or maybe something different. Could the button itself also be responsible for controlling the spacing between the inner elements?

All that said, I'm happy to make changes here.

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: Agreed. I've copied the actual styles into the new styled component for the reset text: https://github.com/WordPress/gutenberg/pull/44260/commits/27cae7cd8764b8e8d3825e36d5058796ea6b03ef

**Review by talldan — APPROVED** [review]
Changes look good to me. Might be good to get a final approval from @mirka, as I'm less familiar with the components package code.

> I've tweaked the text color in the disabled state to match the Saved button as suggested. When hovered

How are you feeling about this, @aaronrobertshaw? It looks very subtle to me, and I'm not sure it makes enough of a difference. I don't mind so much if it's changed back to how it was before. Thanks for trying it.

**Comment by aaronrobertshaw** [comment]
> How are you feeling about this, @aaronrobertshaw? It looks very subtle to me, and I'm not sure it makes enough of a difference. I don't mind so much if it's changed back to how it was before. Thanks for trying it.

I don't have strong feelings about this one. Personally, I'd probably lean a little towards the current state of things. While the difference is very subtle there's at least some form of additional clue that the menu item is disabled.

Perhaps @jasmussen might have some thoughts on the current visuals that might tilt which way we go.

> Might be good to get a final approval from @mirka, as I'm less familiar with the components package code.

No arguments here 🙂

**Comment by jameskoster** [comment]
Might it be better to only display the 'Reset' button when it's possible to perform the action? I'm not sure how much value there is in seeing the disabled button... it could get quite noisy when there are many reset-able defaults.

**Comment by jasmussen** [comment]
`$gray-700` meets contrast requirements to be legible against a white background, but still looks disabled, and as noted there's precedence with the "Saved" indicator. I like Jay's idea of combining it with removing the Reset button. More importantly, now the focus style isn't semi-transparent, which was the primary problem before.

If this one can land and still make it to the backports, I'd just press the green button and look at hiding the Reset button separately.

File: packages/components/src/menu-item/test/index.js
Comment (ciampo) [inline packages/components/src/menu-item/test/index.js:126]: I don't think that the `Icon` text would ever be displayed in the component, regardless of the `suffix` prop being defined or not — in this case, I believe that the component would render something like 

```html
<span class="dashicon dashicons dashicons-Icon"></span>
```

Did you mean to pass `icon={ <span>Icon</span> }` like in the test below?

File: packages/components/src/menu-item/index.js
Comment (ciampo) [reply packages/components/src/menu-item/index.js:76]: Looks good!

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: Yep, thank you!

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: Thank you!

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: If I understand Lena's request correctly, she may have meant that the `margin-left` would be applied by `MenuItem` only when the `suffix` prop is defined. But I'll let her clarify tomorrow when she's back.

> I could see different suffixes desiring no margin or maybe something different. Could the button itself also be responsible for controlling the spacing between the inner elements?

Interesting! What scenarios do you have in mind?

Finally, regardless of where we use `margin-left`, we should make sure if will work well for RTL layouts (ie. potentially wrapping it in a `rtl` call).

---

Potentially, as a follow-up, we could consider refactoring `MenuItem` to use `HStack` and the `gap` prop:

- we could expose the `gap` prop as a way to let consumers tweak spacing around the suffix
- RTL layouts would automatically taken care of (because of `flexbox`)

**Review by ciampo — COMMENTED** [review]
Chiming in as Lena is AFK today — most feedback seems to be addressed apart from https://github.com/WordPress/gutenberg/pull/44260#discussion_r974247615

File: packages/components/src/menu-item/test/index.js
Comment (aaronrobertshaw) [reply packages/components/src/menu-item/test/index.js:126]: 🤦 I did mean to. Thanks. I've fixed this test in https://github.com/WordPress/gutenberg/pull/44260/commits/beba370b36be0c61949d7495bb4231d133549e61

File: packages/components/src/tools-panel/styles.ts
Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: > What scenarios do you have in mind?

My comment was more in the general sense and not trying to assume I'll know what consumers might wish to render in a menu item. I've wrapped the `margin-left` in an `rtl` call for now (https://github.com/WordPress/gutenberg/pull/44260/commits/6ab67b83c3d19b940fcc0982cdb7cc355939b72d) to move this forward.

> Potentially, as a follow-up, we could consider refactoring MenuItem to use HStack and the gap prop:

I like the idea of reworking `MenuItem` to leverage `HStack`. It's along the lines I was thinking suggesting the button could handle the spacing.

File: packages/components/src/menu-item/style.scss
Comment (aaronrobertshaw) [inline packages/components/src/menu-item/style.scss:11]: It looks like the recent changes to reset box-sizing on everything has negatively impacted the approach taken in https://github.com/WordPress/gutenberg/pull/34406 to ensure consistent menu item widths when handling both checked and unchecked items.

Switching box-sizing back in this circumstance smooths things out but it might be worth improving these styles in a follow-up.

**Review by aaronrobertshaw — COMMENTED** [review]
Thank you everyone for the testing, reviews, and feedback 👍 

I think this PR is getting close now. The latest changes include:
- Wrapping the reset label's `margin-left` in an `rtl` call
- Fixing the menu item suffix test
- Omitting the reset label when the default menu item is unchecked i.e. disabled
- Maintaining a consistent menu width after the reset label omission change above

Here's what I'm getting now:

https://user-images.githubusercontent.com/60436221/191461029-e65007df-d6e2-4aa1-85e6-5b32f735683e.mp4





If there is anything else I've missed or needs tweaking before merging. Please let me know 👍

**Comment by jameskoster** [comment]
Isn't it a little strange to focus disabled buttons in the toolspanel menu? 

Screenshot 2022-09-21 at 11 43 21

Could / should the focus skip directly to Height in the screenshot above?

File: packages/components/src/tools-panel/styles.ts
Comment (mirka) [reply packages/components/src/tools-panel/styles.ts]: Yep, you both understood me correctly 🙂 It seems like all the suffix-like bits of our menu items in Gutenberg currently do have left margins, so I'm leaning towards adding a reasonable amount by default to the MenuItem suffix. (Not exactly a blocker though, if we want to get this PR merged.)

In terms of customizable whitespace, we have been trying out "utility wrapper components" as a way to handle custom padding needs cleanly without bloating the main component API. (For example in [InputControl](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-inputcontrol--with-prefix) and [Dropdown](https://wordpress.github.io/gutenberg/?path=/docs/components-dropdown--with-no-padding).) So that is also a possible approach, if the need ever arises. But currently I don't think it's likely that we'll have a no-padding use case here.

**Comment by aaronrobertshaw** [comment]
> Isn't it a little strange to focus disabled buttons in the toolspanel menu?

It is. I think this would be better addressed in a separate PR though so we can land these changes sooner.

File: packages/components/src/tools-panel/styles.ts
Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: _I'll take over from Lena as she's going to be AFK for the next week_

Let's not consider this as blocking, although it would be nice to work on follow-up in which:

- we move the suffix margin to `MenuItem`
- we look into using `HStack` for `MenuMargin`'s layout
- we consider exposing a `gap` prop if necessary, to tweak horizontal spacing inside `MenuMargin`

How does that sound, @aaronrobertshaw  ?

Comment (aaronrobertshaw) [reply packages/components/src/tools-panel/styles.ts]: Thanks @ciampo, that sounds like a plan to me 👍 

The only thing I'd add is we could potentially convert MenuItem to TypeScript if that helps bring it in line with the rest of the components package.

Comment (ciampo) [reply packages/components/src/tools-panel/styles.ts]: That would be great, although it can also be definitely done separately — up to you!

**Pull request merged**
